sw/device/tests/uart_baud_rate_test.c

```diff
@@ -0,0 +1,111 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdint.h>
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_pinmux.h"
+#include "sw/device/lib/dif/dif_uart.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_console.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/lib/testing/test_framework/ottf_utils.h"
+#include "sw/device/lib/testing/uart_testutils.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+static const uint8_t kSendData[] = "UART baud test!";
+static const uint32_t kBaseAddrs[4] = {
+    TOP_EARLGREY_UART0_BASE_ADDR,
+    TOP_EARLGREY_UART1_BASE_ADDR,
+    TOP_EARLGREY_UART2_BASE_ADDR,
+    TOP_EARLGREY_UART3_BASE_ADDR,
+};
+static uint32_t kBauds[7] = {
+    9600, 115200, 230400, 128000, 256000, 1000000, 1500000,
+};
+
+enum {
+  kBaudCountSilicon = 7,
+  // The two highest Bauds won't work at the clock speed we run the FPGAs at.
+  kBaudCountFpga = 5,
+  kTestTimeoutMillis = 500000,
+};
+
+static volatile uint8_t uart_idx = UINT8_MAX;
+static volatile uint32_t baud_rate = UINT32_MAX;
+
+static dif_uart_t uart;
+static dif_pinmux_t pinmux;
+
+OTTF_DEFINE_TEST_CONFIG(.enable_uart_flow_control = true);
+
+// Send all bytes in `kSendData`, and check that they are received via the
+// loopback mechanism.
+static status_t test_uart_baud(void) {
+  CHECK_DIF_OK(dif_uart_configure(
+      &uart, (dif_uart_config_t){
+                 .baudrate = (uint32_t)baud_rate,
+                 .clk_freq_hz = (uint32_t)kClockFreqPeripheralHz,
+                 .parity_enable = kDifToggleDisabled,
+                 .parity = kDifUartParityEven,
+                 .tx_enable = kDifToggleEnabled,
+                 .rx_enable = kDifToggleEnabled,
+             }));
+  OTTF_WAIT_FOR(baud_rate == 0, kTestTimeoutMillis);
```

Comment:
This is a bad way of synchronising with the host and I want to replace it with something nicer